external/avro/src/test/scala/org/apache/spark/sql/avro/AvroSerdeSuite.scala

```diff
@@ -0,0 +1,165 @@
+  }
+
+  test("Fail to convert with deeply nested field type mismatch") {
+    val avro = SchemaBuilder.builder().record("toptest").fields()
+        .name("top").`type`(createNestedAvroSchemaWithFields("foo", _.optionalFloat("bar")))
+        .noDefault().endRecord()
+    val catalyst = new StructType().add("top", CATALYST_STRUCT)
+
+    assertFailedConversionMessage(avro, deserialize = true,
+      "Cannot convert Avro field 'top.foo.bar' to Catalyst field 'top.foo.bar' because schema " +
+        """is incompatible (avroType = "float", sqlType = INT)""",
+      catalyst)
+
+    assertFailedConversionMessage(avro, deserialize = false,
+      "Cannot convert Catalyst field 'top.foo.bar' to Avro field 'top.foo.bar' because schema is " +
+          """incompatible (sqlType = INT, avroType = "float")""",
+      catalyst)
+  }
+
+  test("Fail to convert for serialization with field count mismatch") {
+    val tooManyFields = createAvroSchemaWithTopLevelFields(_.optionalInt("foo").optionalLong("bar"))
+    assertFailedConversionMessage(tooManyFields, deserialize = false,
+      "Avro top-level record schema length (2) " +
+        "doesn't match Catalyst top-level record schema length (1)")
+
+    val tooFewFields = createAvroSchemaWithTopLevelFields(f => f)
+    assertFailedConversionMessage(tooFewFields, deserialize = false,
+      "Avro top-level record schema length (0) " +
+        "doesn't match Catalyst top-level record schema length (1)")
+  }
+
+  /**
+   * Attempt to convert `catalystSchema` to `avroSchema` (or vice-versa if `deserialize` is true),
+   * assert that it fails, and assert that the _cause_ of the thrown exception has a message
+   * matching `expectedCauseMessage`.
+   */
+  private def assertFailedConversionMessage(avroSchema: Schema,
+    deserialize: Boolean,
+    expectedCauseMessage: String,
+    catalystSchema: StructType = CATALYST_STRUCT): Unit = {
+    val e = intercept[IncompatibleSchemaException] {
+      if (deserialize) {
+        new AvroDeserializer(avroSchema, catalystSchema)
+      } else {
+        new AvroSerializer(catalystSchema, avroSchema, false)
+      }
+    }
+    val expectMsg = if (deserialize) {
+      s"Cannot convert Avro type $avroSchema to Catalyst type ${catalystSchema.sql}."
+    } else {
+      s"Cannot convert Catalyst type ${catalystSchema.sql} to Avro type $avroSchema."
+    }
+    assert(e.getMessage === expectMsg)
+    assert(e.getCause.getMessage === expectedCauseMessage)
+  }
+}
+
+
+object AvroSerdeSuite {
+
+  private val CATALYST_STRUCT = new StructType()
+      .add("foo", new StructType().add("bar", IntegerType))
+
+  /**
+   * Convenience method to create a top-level Avro schema with a single nested record
+   * (at field name `nestedRecordFieldName`) which has fields as defined by those set
+   * on the field assembler using `f`.
+   */
+  private def createNestedAvroSchemaWithFields(
+    nestedRecordFieldName: String,
```

Comment:
nit: 4 spaces indentation for parameters.

Reply:
Argh. This time I had the good sense to audit for any more bad instances and found one additional spot. Thank you.